man/man1/mkgrnlib.1
.TH mkgrnlib 1 "27 April 2023" "MTINV Version 4.0.1" "MTINV Toolkit"

.SH NAME 
mkgrnlib  \- Makes the Green's function library with multiple source depths for a single source-to-receiver station 

.SH SYNOPSIS
mkgrnlib
stnm=(string)
net=(string)
loc=(string)
velmod=(string)
modeldb=(string)
stadb=(string)
[ zrange=(float,float,float) or z_depth_km=(float) or nz=(int) z_multiple=(float,float,...)]
evla=(float)
evlo=(float)
dt=(float)
nt=(integer)
fmax=(float)
t0=(float)
rdev=(float)
damp=(float)
kmax=(integer)
eps=(float)
smin=(float)
[no]dump
[no]verbose
.sp
or... 
.sp
mkgrnlib par=model.par

.SH DESCRIPTION
.B mkgrnlib 
.R computes a Green's function (GF) library (*.glib) over a range of source depths for a single source-to-receiver pair. 
A Green's function is an impulse response of the Earth 
propagation differential operator on a domain with specified inital and boundary conditions.  
Green's functions are computed using the frequency-wavenumber reflectivity method (Zeng and Anderson, 1995). 
We use the 3 fundamental faulting orientations (SS-strike-slip, DS-45deg dip slip, DD-vertical dip) and Isotropic component (EP)
This results in 10 Green's functions per station and depth:  
(ZSS, ZDS, ZDD, ZEP, RSS, RDS, RDD, REP, TSS, TDS).
The first letter is the component (Z, R, T) Z-vertical, R-radial, T-transverse or tangential components. 
This application reads in a parameter file and/or command line arguments for the 
1D earth velocity model, receiver coordinates, and source coordinates.
The model file is in simple ASCII column format (see below). 
The station latitude and longitude are stored in a file (rdseed.stations) provided by RDSEED (now depricated).  
See sac2gmtmap to create an eqivalent file using the -rdseed.station flag.  

.SH REQUIRED PARAMETERS

.TP
.B stnm={string}
Station Code, required cannot be NULL. The station code must be in the same capitalization and spelling as typed in column one of the station database file, see \fIstadb\fP parameter. The station code has an eight character length limit.

.TP
.B net={string}
Network Code, required cannot be NULL (see www.FDSN.org). The network code must be in the same capitalization and spelling as
typed in column 2 of the the station database file, see \fIstadb\fP parameter. The network code
has an eight character length limit.

.TP
.B loc={string}
Location Code, required (for null use empty quotes i.e., loc=""). 
When not NULL, the location code is usually a two digit number, e.g., 00, 10, 20, 01, 02, 60, etc...

.TP
.B velmod={string}
The file name of the velocity model without the .mod extension.  The path is defined by \fImodeldb\fP (see below).
[required no default].

.TP
.B zrange={float},{float},{float}
The range of source depths to compute in the Green's function library.  
The zrange is a comma seperated vector represented by 
the minimum depth, increment, and maximum depth values in (kilometers). The alternative is to
supply a single depth using \fIz_depth_km=(float)\fP parameter.  If this is set then zrange is skipped. 

.TP
.B nz=(integer) z_multiple=(float,float,...)
The range of any source depths to compute in the Green's function library. For example, nz=6 z_multiple=1,5,8,14,32,50.
If this is set then \fIzrange\fP and \fIz_depth_km\fP are skipped. 

.TP
.B evla={float}
The earthquake latitude in decimal degrees format [required no default].

.TP
.B evlo={float}
The earthquake longitude in decimal degrees format [required no default].

.TP
.B dt={float}
The Green's function sampling rate in seconds per sample  [required no default].

.TP
.B nt={integer}
The number of points for the Green's functions.  Must be a power of 2 (i.e., 64, 128, 256, 512, 1024, 2048(maximum))
[required no default].

.TP
.B eps={float}
Error tolerance.  Typically values of \fIeps\fP=0.001 to \fIeps\fP=0.000001 are adequate.  When \fIkmax\fP or \fIfmax\fP are not set 
properly mkgrnlib will exit and ask to reset these values.  This feature can be controlled by \fIeps\fP and \fIsmin\fP.
[required no default].

.TP
.B smin={float}
Error tolerance.  Typically values of \fIsmin\fP=0.001 to \fIsmin\fP=0.000001 are adequate.  When \fIkmax\fP or \fIfmax\fP are not set properly mkgrnlib will exit and ask to reset these values.  This feature can be controlled by \fIeps\fP and \fIsmin\fP.
[required no default].

.TP
.B modeldb={string}
The path to the model database directory.
[required no default].

.TP
.B stadb={string}
The path and file name to the station database.  The file is in rdseed format.
[required no default].

.SH OPTIONAL PARAMETERS

.TP
.B par={string}
The name of the parameter file [optional no default].  Allows all or some of the 
parameters to be defined in a file rather than
on the command line.  This is ideal for static parameters, for example,
\fImodeldb\fP and \fIstadb\fP, which may not change for each new run.

.TP
.B fmax={floating point number}
The Green's function maximum frequency content.  The absolute maximum is the Nyquist frequency.  
For moment tensor inversion, the
Nyquist frequency is too high for fmax and can be lowered without loss of information.  This speeds calculation time
significantly.  Typically \fIfmax\fP can be set to 0.5 seconds when inverting 100 to 10 sec period waves
[optional, default is Nyquist frequency=1/(2*dt)].
                                                                                                                                                 
.TP
.B t0={floating point number}
The time of the first sample in the Green's function, (\fIt0\fP=0 is the earthquake origin time) 
[optional, default is \fIt0\fP=0.0]
                                                                                                                                                 
.TP
.B redv={floating point number}
The reduction velocity in km/sec [optional, default is \fIredv\fP=-1.0 (no reduction velocity)].  To avoid
wasting time computing Green's function for times before the first arrival particularly when the origin-time
is used as the time of the first sample at large distances.  The reduction velocity can move the time of the
first sample to a later time based on a horizontal apparent velocity.

.TP
.B damp={floating point number}
The value damp is for the exp(-damp*(pi/twin)*t) exponential function for dampening the
rap around effects in the time domain. 
When the length of the time history is too short, then the seismic surface waves may
rap around to the beginning of the Green's function
because of the cyclic nature of the FFT [optional, default off \fIdamp\fP=1.0].
                                                                                                                                                 
.TP
.B kmax={integer number}
Similar to fmax but for maximum spatial frequency (k=wave number).  Warning low fmax and kmax values may led to instability and inaccuracies
in the calculations.  [optional, default is kmax=10000000].  Typically kmax=20000 is adequate for most moment tensor inversions.
                                                                                                                                                 
.TP
.B verbose={integer=1 or 0}
Verbosy output for debugging is \fIverbose\fP=1 and for no verbosy output is \fIverbose\fP=0 [optional, default no verbose=0].

.TP
.B dump={integer=1 or 0}
Write out "Dump" the Green's functions as SAC formatted binary files [optional, default no dump=0].

.SH VELOCITY MODEL FILE FORMAT
###
.br
### Western US model Ritsema and Lay 1995 JGR.
.br
###
.br
### thick Vp     Qp      Vs    Qs      rho
.br
###  (km) (km/s)        (km/s)         (g/cc)
.br
   4.00  4.52   500.00  2.61   250.00  2.39
.br
  28.00  6.21   500.00  3.59   250.00  2.76
.br
  20.00  7.73  1000.00  4.34   500.00  3.22
.br
 700.00  7.64  1000.00  4.29   500.00  3.19
.br
.sp

.SH Example entries in the station database file
GSC CI +35.301800 -116.805700  +954.0 "BHE BHN BHZ" "Goldstone, California, USA" 1990,220,00:00:00 2599,365,23:59:59
.br
BAR CI +32.680100 -116.672200  +496.0 "BHE BHN BHZ" "Barrett, California, USA" 1992,275,00:00:00 2599,365,23:59:59
.br
.sp
only columns 1, 2, 3, 4, and 5 are read by mkgrnlib. 

.SH EXAMPLE
To calculate the Green's functions of an earthquake in Anza, California (latitude=33.529, longitude=-116.573) 
at the UC Berkeley digital seismograph 
station Columbia (CMB.BK) within the central Sierra Nevada, California,
for depths of 2 to 22 km in 2 km increments (i.e., 2,4,6,8,10,12,14,16,18,20,and 22 km).  
The Western U.S. velocity model is used and the time histories will be computed up to 0.5 Hz,
at a sampling rate of 0.2 samples per second and a total of 1024 points (204.8 seconds).  
.br
.sp
\fBmkgrnlib\fP
evla=33.529 
evlo=-116.573 
zrange=2,2,22 
velmod=wus \\
.br
  dt=0.2
nt=1024
fmax=0.5
eps=0.0005
smin=0.0005 \\
.br
  modeldb=/Users/ichinose/mtinv.v0.9/modeldb/ \\
.br
  stadb=/Users/ichinose/mtinv.v0.9/stadb/station_database.txt

.SH "SEE ALSO"
.IR sac2gmtmap (1),
.IR multithread_mkgrnlib (1),
.IR glib2inv (1),
.IR sacdata2inv (1),
.IR mtinv (1),
.IR grnlib2sac (1),
.IR grn2Mxy (1)


src/mkgrnlib.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "../include/mt.h" /** global datatype and structure declarations **/

char progname[128];

/*** computes Greens functions just one station at a time ***/
/*** see ../misc/multithread_makeglib/multithread_mkgrnlib.c uses fork-multiprocessing ***/

int main( int ac, char **av )
{
	extern Greens grn_;
	int iz;
	float *z;
	Depth_Info depth_info;
	FILE *fp;
	int verbose=0, idump=0, ista;
	void Usage_Print(void);
	int it;  /*** this version computes the rigidity here ***/
	float cm2km, UnitArea, rigidity, slip4pi;
	cm2km = 1.0E+05;
	UnitArea = 1 * cm2km * cm2km;

/* see include/mt.h Mo = math.pow( 10.0, 1.5*(Mw+10.73) ) = 1.2445146117713818e+16; Reference Mw = 0.0 */

	int isrclayer;
	VelMod vmint;
	float dlaykm  = 1.0;

	void greensf_( int * );

	/*** modified in version 4.0, added string char *wavetype = "Surf/Pnl" or "Rotational" ***/
	void wrtgrn2sac( Greens *g, int ista, char *wavetype, char *fmt, int make_output_dirs );

	void getparameters( int, char **, Greens *, Depth_Info *, int *, int * );

/*** from modsubs.c ***/
        void create_mod( VelMod * );
        void print_mod0( VelMod * );
        void print_mod1( FILE *, VelMod * );
        void compute_rigidity( Greens *, int );

/*****************************/
/*** from rayp/rayp_subs.c ***/
/*****************************/
        void interpolate_model( VelMod *, VelMod *, float );
        void calc_1D_tt( float, float, float *, float *, float *, int *, VelMod *, int );
        void earth_flattening_transformation( VelMod * );
        float calc_takeoff_angle( float, float, float, float );

/*** misc ***/
	char pathname[128];
	char *shorten_path( char *pathname, char *filename );

/*****************************/
/*** start program main    ***/
/*****************************/
	strcpy( pathname, av[0] );

	shorten_path( pathname, progname );

	fprintf( stdout, "%s: %s: %s: STDOUT: Version=%s Release Date=%s exec full path=%s\n", 
		progname, 
		__FILE__,
		__func__,
		Version_Label,
		Version_Date,
		pathname );

	if( ac <= 1 ) Usage_Print();

/*******************************************************/
/*** get input parameters from file and command line ***/
/*******************************************************/
	depth_info.zvec = calloc(200,sizeof(float));
	getparameters( ac, av, &grn_, &depth_info, &verbose, &idump );

/**********************************************************/
/*** make an interpolated version of the original model ***/
/*** clones the original velocity model for use to calc ***/
/*** 1D travel-times and ray-parameters                 ***/
/**********************************************************/
	interpolate_model( &(grn_.v), &vmint, dlaykm );
	if(verbose) print_mod0( &vmint );
                                                                                                                                 
/**********************************************************/
/*** Earth Flattening Transformation                    ***/
/*** EFT is applied only onto interpolated velocity     ***/
/*** model from above and use to calc 1D travel-times   ***/
/**********************************************************/
	earth_flattening_transformation( &vmint );
	if(verbose)print_mod0( &vmint );

/*******************************************************/
/*** set the depth vector                            ***/
/*******************************************************/
	if( depth_info.type == 0 || depth_info.type == 1 )
	{
	  z = (float *) calloc( depth_info.nz, sizeof(float) );
	  if(verbose) printf("%s: \t\t", progname );
	  for( iz=0; iz < depth_info.nz; iz++ )
	  {
		z[iz] = depth_info.zmin + ( iz * depth_info.zinc );
		if(verbose)
		{
			if( iz == depth_info.nz - 1 ) printf("%g\n", z[iz] );
			else printf("%g ", z[iz] );
		}
	  }
	}
	else if( depth_info.type==2 )
	{
	  z = (float *) calloc( depth_info.nz, sizeof(float) );
	  if(verbose) printf("%s: \t\t", progname );
	  for( iz=0; iz < depth_info.nz; iz++ )
	  {
		z[iz] = depth_info.zvec[iz];
	        if(verbose)
		{
			if( iz == depth_info.nz - 1 ) printf("%g\n", z[iz] );
			else printf("%g ", z[iz] );
	 	}
	  }
	}

/*******************************************************/
/*** open the file and write the source depth        ***/
/*** references in the header of the output file     ***/
/*******************************************************/
	if( (fp = fopen(grn_.filename,"wb")) == NULL )
	{
		printf("%s: Fatal Error, cannot open file %s\n", 
			progname, grn_.filename );
		exit(-1);
	}
	fwrite(&(depth_info.nz),sizeof(int),1,fp);
	fwrite(&z[0],depth_info.nz*sizeof(float),1,fp);
	fflush(fp);

/*******************************************************/
/*** loop over depth                                 ***/
/*******************************************************/
	for( iz=0; iz< depth_info.nz; iz++ )
	{

	/*******************************************************/
	/*** compute the rigidity at the source for the      ***/
	/*** source depth's velocity and density             ***/
	/*******************************************************/
		if( verbose ) printf("%s: iz=%d z=%g\n", progname, iz, z[iz] );
		grn_.evdp = z[iz];
		compute_rigidity( &grn_, verbose );

	/********************************************************/
	/*** compute the Green's functions Ref:               ***/
	/*** Zeng and Anderson (1995) Bull. Seismol. Soc. Am. ***/
	/********************************************************/
		greensf_( &verbose );

	/***********************************************************/
	/*** dump part of the structure for debugging            ***/
	/***********************************************************/
		if(verbose)
		{
		  fprintf(stdout, "%s: filename=%s stnm=%s net=%s loc=%s stla=%f stlo=%f ",
		    progname, grn_.filename, grn_.stnm, grn_.net, grn_.loc, grn_.stla, grn_.stlo );

		  fprintf(stdout, "evla=%f evlo=%f evdp=%f rdist=%g az=%f baz=%f ",
		    grn_.evla, grn_.evlo, grn_.evdp, grn_.rdist, grn_.az, grn_.baz );

		  fprintf(stdout, "t0=%f dt=%f fmax=%f damp=%f eps=%f smin=%f\n",
		    grn_.t0, grn_.dt, grn_.fmax, grn_.damp, grn_.eps, grn_.smin );
		}

		fprintf(stdout, 
	"%s: %s: %s: %s evdp=%6.2f rdist=%8.2f az=%3.0f nt=%4d dt=%5.3f fmax=%5.2f t0=%g twin=%g ",
			progname,
			__FILE__,
			__func__,
			grn_.filename,
			grn_.evdp,
			grn_.rdist,
			grn_.az,
			grn_.nt,
			grn_.dt,
			grn_.fmax,
			grn_.t0,
			grn_.twin );

	/**************************/
	/*** apply the rigidity ***/
	/**************************/
		rigidity = grn_.rigidity;
		slip4pi =  (base_moment/(rigidity * UnitArea))/(4*M_PI);
		for( it = 0; it < grn_.nt; it++ )
		{
			grn_.g.rss[it] *= slip4pi;
			grn_.g.rds[it] *= slip4pi;
			grn_.g.rdd[it] *= slip4pi;
			grn_.g.rep[it] *= slip4pi;
			grn_.g.zss[it] *= slip4pi;
			grn_.g.zds[it] *= slip4pi;
			grn_.g.zdd[it] *= slip4pi;
			grn_.g.zep[it] *= slip4pi;
			grn_.g.tss[it] *= slip4pi;
			grn_.g.tds[it] *= slip4pi;
		}

	/************************************/
	/*** introduce reduction velocity ***/
	/************************************/
		grn_.ts0	= grn_.t0;
		grn_.tstart	= grn_.ts0;
		grn_.tend	= grn_.ts0 + grn_.dt * grn_.nt;

	/*********************************************************************/
	/*** calculate travel time, ray parameter, and ray bottoming depth ***/
	/*********************************************************************/
		verbose = 0;
		calc_1D_tt( grn_.rdist, 
			grn_.evdp, 
			&(grn_.Prayparameter), 
			&(grn_.Pttime), 
			&(grn_.Praybottom), 
			&isrclayer, 
			&vmint, 
			verbose );

		grn_.Ptakeoff = calc_takeoff_angle( 
			vmint.vp[isrclayer], 
			grn_.Prayparameter, 
			grn_.rdist, 
			grn_.evdp );

		  fprintf( stdout, " p=%6.4f rb=%6.1f tt=%7.2f toa=%5.1f\n",
                    grn_.Prayparameter,
                    grn_.Praybottom,
                    grn_.Pttime,
                    grn_.Ptakeoff );

	/***********************************************************/
	/*** write out greens functions as SAC files for testing ***/
	/*** this code only computes translational GFs not RotGF ***/
	/***********************************************************/
		ista = 0;
		if( idump ) wrtgrn2sac( &grn_, ista, "Surf/Pnl", "\0", 0 /* int make_output_dirs */ );

	/**************************************************************/
	/*** write out binary greens functions library for glib2inv ***/
	/**************************************************************/
		fwrite(&grn_,sizeof(Greens),1,fp);
	}

/****************/
/*** clean up ***/
/****************/
	fclose(fp);
	free(z);
	printf("%s: program finished. Bye-Bye!\n\n\n", progname );
	exit(0);
}

void getparameters( int ac, char **av, Greens *g, 
		Depth_Info *depth_info, int *verbose, int *idump )
{
	double drdist,daz,dbaz;
	float nyquist_frequency;
	char stadb_filename[128];

	int distaz( double, double, double, double, double *, double *, double * );
	int setpar( int, char ** );
	int getpar( char *, char *, void * );
	int mstpar( char *, char *, void * );
	void endpar();
	void create_mod( VelMod * );
	void print_mod0( VelMod * );
	void getsta( char *, Greens *, int * );

/***depth***/
	float single_src_depth_km = -99; /*** type=0 ***/
	int   nz_multiple_irregular = 0; /*** type=2 ***/
	/*** temp space for getpar() ***/
	float zvec[3];
	char  zstr[8];
	int   iz;

/***************************** begin **********************************/

	setpar(ac,av);
/*************************/
/**** output settings ****/
/*************************/
	getpar("verbose", "b", verbose );
	getpar("dump", "b", idump );

/**************************************/
/**** set the depth or depth range ****/
/**************************************/
	getpar( "z_depth_km", "f", &single_src_depth_km );
	getpar( "nz", "d", &nz_multiple_irregular );

	if( single_src_depth_km == -99 && nz_multiple_irregular == 0 )
	{
	 mstpar("zrange", "vf[3]", &zvec );
	 depth_info->type = 1;
	 depth_info->zmin = zvec[0];
	 depth_info->zinc = zvec[1];
	 depth_info->zmax = zvec[2];
	 depth_info->nz   = ((depth_info->zmax-depth_info->zmin)/depth_info->zinc) + 1;
	}
	else if( single_src_depth_km == -99 && nz_multiple_irregular > 0 )
	{
	 depth_info->type = 2;
	 depth_info->nz   = nz_multiple_irregular;
	 depth_info->zvec = (float *)realloc( depth_info->zvec, nz_multiple_irregular * sizeof(float) );
	 sprintf( zstr, "vf[%d]", nz_multiple_irregular );
	 mstpar( "z_multiple", zstr, depth_info->zvec );
	 for( iz = 0; iz < nz_multiple_irregular; iz++ )
	 {
		fprintf( stdout,  "iz=%d z_multiple_irregular=%g\n", iz, depth_info->zvec[iz] );
	 }
	}
	else if( single_src_depth_km != -99 && nz_multiple_irregular == 0 ) /* kludge for single src depth */
	{
	 depth_info->type = 0;
	 depth_info->nz   = 1;
	 depth_info->zmin = single_src_depth_km;
	 depth_info->zinc = single_src_depth_km*2;
	 depth_info->zmax = single_src_depth_km;
	}
	else
	{
		fprintf(stderr,"%s: %s: %s: ERROR! in depth_info\n", progname, __FILE__, __func__ );
		exit(-1);
	}

	if( *verbose )
	{
	  printf("%s: Compute Greens Functions for:\n", progname );
	  printf("%s: \t\t zmin=%g zinc=%g zmax=%g nz=%d\n",
	    progname, depth_info->zmin, depth_info->zinc, 
		depth_info->zmax, depth_info->nz );
	}

/************************************************************************/
/*** get the station location information, later get this information ***/
/*** from station geometry file including station and network         ***/
/************************************************************************/
	mstpar("stnm", "s", &(g->stnm) );
	mstpar("net",  "s", &(g->net) );
	mstpar("loc",  "s", &(g->loc) );
	mstpar("stadb", "s", stadb_filename );

/*************************************/
/*** source latitude and longitude ***/
/*************************************/
	mstpar("evla", "f", &(g->evla) );
	mstpar("evlo", "f", &(g->evlo) );

/*** important stuff ***/
	mstpar("nt",	"d", &(g->nt) );
	mstpar("dt",	"f", &(g->dt) );
	g->twin = g->nt * g->dt;

/*** f-k stuff ***/
	mstpar("eps",	"f", &(g->eps) );
	mstpar("smin",	"f", &(g->smin) );
	g->damp = 1.0;
	getpar("damp",  "f", &(g->damp) );
	g->kmax = 10000000;
	getpar("kmax",  "d", &(g->kmax) );

/**************************/
/*** reduction velocity ***/
/**************************/
	g->redv  = -1.;
	getpar("redv",  "f", &(g->redv) );
	g->t0 = 0;
	getpar("t0", "f", &(g->t0) );

/***************************************************************/
/*** leave off the .mod in filename append only when opening ***/
/***************************************************************/
	mstpar("modeldb", "s", &(g->v.modpath) );
	mstpar("velmod", "s", &(g->v.modfile) );

/******************************************************************************/
/*** default fmax is the nyquist frequency unless otherwise set to be lower ***/
/******************************************************************************/
	nyquist_frequency = 1/(2 * g->dt);
	getpar("fmax", "f", &(g->fmax) );
	if( g->fmax > nyquist_frequency ) g->fmax = nyquist_frequency;

	g->twin = g->nt * g->dt;

	endpar();

/***********************************/
/*** get the station information ***/
/***********************************/
	g->stel = 0;
	getsta( stadb_filename, g, verbose );
	fprintf( stdout, "%s: %s: %s: Net.Sta.Loc=(%s.%s.%s) stla=%g stlo=%g stel=%g evla=%g evlo=%g\n",
		progname, __FILE__, __func__,
		g->net, g->stnm, g->loc, g->stla, g->stlo, g->stel,
		g->evla, g->evlo );

	if( distaz( (double)g->evla, (double)g->evlo, 
		(double)g->stla, (double)g->stlo, &drdist, &daz, &dbaz ) == 0 )
	{
		g->rdist = (float)drdist;
		g->az    = (float)daz;
		g->baz   = (float)dbaz;
		fprintf( stdout, 
			"%s: %s: %s: distazs(r=%g, az=%g, baz=%g)\n", 
			progname, __FILE__, __func__, 
			g->rdist, g->az, g->baz );
	}
	else
	{
		fprintf( stdout, "%s: %s: %s: distaz: returned a fatal error\n", 
			progname, __FILE__, __func__ );
		exit(-1);
	}

/********************************************************************/
/*** use the reduction velocity (km/sec) to set the t0 time shift ***/
/********************************************************************/
	if ( g->redv > 0 ) 
	{
	  g->t0     = g->rdist/g->redv;
	  g->tstart = g->ts0 = g->t0;
	  g->tend   = g->tstart + g->twin;

	  fprintf( stdout, "%s: %s: %s: dist=%g reduction velocity=%g(km/sec) t0=%g ",
		progname, __FILE__, __func__, g->rdist, g->redv, g->t0 );

	  fprintf( stdout, "tstart=%g tend=%g twin=%g\n",
		g->tstart, g->tend, g->twin ); 
	}

/*************************************************/
/*** set the green's function file name        ***/
/*** changes 2019/Dec G. Ichinose includes loc ***/
/*** net.sta.loc.model                         ***/
/*************************************************/
			/* 256 -> 8 + 1 + 8 + 1 + 8 + 256 = 285 */ 
/*** 
clang -I../include -c mkgrnlib.c -o mkgrnlib.o
mkgrnlib.c:449:2: warning: 'snprintf' will always overflow; destination buffer has size 256, but size argument is 285 [-Wbuiltin-memcpy-chk-size]
        snprintf( g->filename, 285,
***/
	snprintf( g->filename, 256,
		"%s.%s.%s.%s.glib",
		g->net,
		g->stnm,
		g->loc,
		g->v.modfile );

       	fprintf( stdout, "%s: %s: %s: Writing glib file to %s\n", 
		progname, __FILE__, __func__, g->filename );

/*********************************************/
/*** set the velocity model data structure ***/
/*********************************************/
	create_mod( &(g->v) );
	/** if( *verbose ) print_mod0( &(g->v) ); **/
	print_mod0( &(g->v) );
}

void Usage_Print()
{
	fprintf(stderr, "\n");
	fprintf(stderr, " USAGE: %s par=(string)\n", progname );
	fprintf(stderr, "\n"); 
	fprintf(stderr, " USAGE: %s stnm=(string) net=(string) loc=(string) velmod=(string)  modeldb=(string) \\ \n", progname );
	fprintf(stderr, "                 stadb=(string) zrange=(float,float,float) evla=(float) evlo=(float) dt=(float) \\ \n" );
	fprintf(stderr, "                 nt=(integer) fmax=(float) t0=(float) rdev=(float) damp=(float) kmax=(integer) \\ \n" );
	fprintf(stderr, "                 eps=(float) smin=(float) [no]dump [no]verbose\n" );
	fprintf(stderr, "\n"); 
	fprintf(stderr, "REQUIRED PARAMETERS:\n" );
	fprintf(stderr, "\t stnm=(string)        Station Code, required cannot be NULL\n" );
	fprintf(stderr, "\t net=(string)         Network Code, required cannot be NULL\n" );
	fprintf(stderr, "\t loc=(string)         Location Code, required for null use empty quotes i.e., loc=\"\")\n" );
	fprintf(stderr, "\t velmod=(string)      Velocity model base name without .mod extenstion\n" );
	fprintf(stderr, "\t modeldb=(string)     Directory path to the velocity model files\n" );
	fprintf(stderr, "\t stadb=(string)       Directory path and filename of the station locations file\n" );
	fprintf(stderr, "\n"); 
	fprintf(stderr, "\t There are three ways to do source depth for Green's function library:\n" );
	fprintf(stderr, "\t zrange=(flt,flt,flt) Range of source depths in km, e.g., zrange=starting,increment,stop \n");
	fprintf(stderr, "\t z_depth_km=(float)   Single source depth in km [optional: if set then zrange skipped]\n");
	fprintf(stderr, "\t nz=(int) z_multiple=(float,float,...)  multiple depths any set or order\n" );
	fprintf(stderr, "\t                           [optional: if nz!= then zrange and z_depth_km skipped]\n" );
	fprintf(stderr, "\n"); 
	fprintf(stderr, "\t evla=(float)         Source latitude, in decimal degrees\n" );
	fprintf(stderr, "\t evlo=(float)         Source longitude, in decimal degrees\n" );
	fprintf(stderr, "\t dt=(float)           Green's function sample rate (samples per seconds)\n" );
	fprintf(stderr, "\t nt=(integer)         Green's function number points (power of 2)\n" );
	fprintf(stderr, "\t fmax=(float)         Green's function maximum frequency (max is Nyquist)\n" );
	fprintf(stderr, "\n"); 
	fprintf(stderr, "\t There are two ways to do starting time for Green's function synthetic origin-time:\n" );
	fprintf(stderr, "\t t0=(float)           Green's function starting time in seconds\n" );
	fprintf(stderr, "\t rdev=(float)         Green's function starting from a reduction velocity (km/sec)\n");
	fprintf(stderr, "\t                                     t0_sec = ( distance_km / reduction_velocity )\n");
	fprintf(stderr, "\n"); 
	fprintf(stderr, "\t damp=(float)         Green's function damping (1 is OK)\n" );
	fprintf(stderr, "\t kmax=(integer)       Green's function maximum wavenumber\n" );
	fprintf(stderr, "\t eps=(float)          Green's function accuracy tolorance parameter 1\n" );
	fprintf(stderr, "\t smin=(float)         Green's function accuracy tolorance parameter 2\n" );
	fprintf(stderr, "\n"); 
	fprintf(stderr, "OPTIONAL PARAMETERS:\n" );
	fprintf(stderr, "\t [no]verbose          Be verbosy [default off]\n" );
	fprintf(stderr, "\t [no]dump             Write out the GF synthetics in SAC files [default off]\n" );
	fprintf(stderr, "\n" );
	fprintf(stderr, "EXAMPLE:\n" );
	fprintf(stderr, "\t %s stnm=CCM net=IU loc=00 velmod=cus modeldb=${MTINV_PATH}/data/modeldb \\ \n", progname );
	fprintf(stderr, "\t          zrange=1,1,25 evla=36.9077 evlo=-90.543 dt=0.15 nt=2048 fmax=0.4 t0=0. redv=18. \\ \n" );
	fprintf(stderr, "\t          damp=1. kmax=20000 eps=0.0005 smin=0.0005 stadb=../Data/rdseed.stations noverbose nodump\n" );
	fprintf(stderr, "\n" );
	fprintf(stderr, "\t Outputs a ~10MB Green's function library file \"IU.CCM.00.cus.glib\" in current directory\n" );
	fprintf(stderr, "\n" );
}
